### PR TITLE
WIP: widget reinforcement learning

### PR DESCRIPTION
**Pull request by mishig25**
# TLDR: automatically play `.mp4` file in hub repos whose library=stable-baselines3. 

## Current approach:
1. User needs to push .mp4 file to the hub:
```python
push_to_hub(repo_id = "ThomasSimonini/ppo-LunarLander-v2",
           filename = "output.mp4", 
           commit_message = "Added LunarLander-v2 video")
```
2. Then, update the README.md as 
```html
<video src="https://huggingface.co/ThomasSimonini/ppo-LunarLander-v2/resolve/main/output.mp4" controls autoplay loop/>
```
see example [here](https://huggingface.co/ThomasSimonini/ppo-LunarLander-v2/blob/main/README.md)

## Proposed approach:
1. On `push_to_hub` python impl,  add logic where if filename.endsWith('mp4'), automatically, update the model card metadata:
```yaml
widget:
    recording:
        - output.mp4
```
2. By using this updated model card metadata, play the video files in the widget section of model page
<video src="https://user-images.githubusercontent.com/11827707/152698468-a5acc062-3f06-4b17-b5c5-a27bffeee912.mov" controls autoplay loop/>

### CODE REVIEW

**Comment by julien-c** [comment]
the widget and the publishing workflow are kinda different issues no?

- I like the widget
- i like the metadata proposal (could support multiple videos, with maybe a way to go to next one)

- not sure about the automated workflow "On push_to_hub python impl, add logic where if filename.endsWith('mp4'), automatically, update the model card metadata" => it should be the training script/library that does that no? or alternatively, on moon-landing side, if model == RL && there's a video.mp4 inside the repo, play it (not sure about option 2, i think option 1 is better)

**Comment by mishig25** [comment]
>  it should be the training script/library that does that no?

afaik, there is no connection between RL library (stable-baslines3 for example) training loop & [huggingface_sb3](https://github.com/huggingface/huggingface_sb3). Therefore, I don't think it is possible to do so (if I'm not mistkane, cc @simoninithomas)

Which leaves us two options:
1. moon-landing side as you desctibe above
2. the initial approach I described. Isn't there a way we can change model card frontmatter when mp4 files gets `push_to_hub`-ed?

**Comment by simoninithomas** [comment]
We talked yesterday with Mishig about **how we can use it with SB3.**

We have two solutions:

1️⃣ Adding a method into push_to_hub() in huggingface_sb3 that generates a video of the agent performing. The video is pushed to hub to and the widget detect if there is a output.mp4 in the repo.
➕: It's easier because the user has the environment installed, so the video generation is smooth.
➖: It implies to implement this system for each library.

2️⃣ Doing this process in the hub itself. When someone push to hub a model, we generate a video of the agent performing in the environment
➕: It's more scalable
➖: It implies to be able to handle every possible environment.

I think personally, that the first solution for now is the best. In the future (like in a year), with the project of hosted environments on the Cloud (Edward's project) we might move towards the second solution. But we wanted to have your position @julien-c and @osanseviero

**Comment by osanseviero** [comment]
Yes, option 1 sounds like the right way.

It's the responsibility of the libraries to create the video and update the metadata - doing this in the Hub side won't scale easily to different libraries (lack of flexibility, doing things in a non transparent way, etc)

**Comment by julien-c** [comment]
Yes fully agree.

The only moment when 2️⃣ will be doable is when we build some kind of repo-centric, scriptable "Hugging Face Actions" workflows but that's a long way into the future =)

**Comment by Wauplin** [comment]
Just read the thread here. 
I'll close this PR since it seems that you agreed on implementing it in a separate repo + has been continued in https://github.com/huggingface/hub-docs/pull/75.
